src/target/riscv/debug_register_printers.c

```diff
@@ -0,0 +1,79 @@
+#include <stdio.h>
+#include <inttypes.h>
+#include <assert.h>
+#include <stdarg.h>
+
+#include "debug_register_printers.h"
+
+unsigned int get_len_or_sprintf(char *buf, unsigned int curr, const char *format, ...)
```

Comment:
This makes things a lot more complex, just so we can avoid allocating an extra 30 or so bytes on the stack. It's simpler and safer to let the user of this file allocate a static buffer and to pass that in, along with the size of the buffer. Then use snprintf to not write any bytes outside that buffer. If the caller is really concerned that it might not pass enough buffer, then they can check the length of the resulting string (which must still be null-terminated).

Reply:
Please, consider the following:
* The length of the buffer required can differ a lot, depending on the register.
```
0x400382 { version=2, confstrptrvalid=invalid, hasresethaltreq=0, authbusy=ready, authenticated=true, anyhalted=1, allhalted=1, anyrunning=0, allrunning=0, anyunavail=0, allunavail=0, anynonexistent=0, allnonexistent=0, anyresumeack=0, allresumeack=0, anyhavereset=0, allhavereset=0, impebreak=1, stickyunavail=current, ndmresetpending=false, }

0x321009 { regno=0x1009, write=arg0, transfer=enabled, postexec=disabled, aarpostincrement=disabled, aarsize=64bit, cmdtype=0, }
```
So it is not `30` and more like `O(<average_n_fields> * <average_field_name_l + average_field_value_name_l>)`.

There is an option to calculate the maximum required length of the buffer during code generation, but this value depends on the format of the output, so, to be safe, it would be required to specify the format in `debug_defines.h`.

 > If the caller is really concerned that it might not pass enough buffer, then they can check the length of the resulting string (which must still be null-terminated).

This requires the user to perform additional actions (apart from allocation and the call to the function), so the resulting client code is not as simple as it could be.

To conclude:
```
char buf[riscv_debug_reg_to_s(NULL, reg, context, value) + 1];
riscv_debug_reg_to_s(buf, reg, context, value);
```
vs
```
char buf[<SOME_ARBITRARY_CONSTANT>];
riscv_debug_reg_to_s(buf, <SOME_ARBITRARY_CONSTANT>, reg, context, value);
```
IMHO the second approach is not better.

As a compromise, it is possible to add the length of the buffer to `riscv_debug_reg_to_s`'s arguments (with corresponding checks inside). However, this would require the user to write more code (if one does not wants to use the `<SOME_ARBITRARY_CONSTANT>`):
```
size_t buf_len = riscv_debug_reg_to_s(NULL, reg, context, value) + 1;
char buf[buf_len];
riscv_debug_reg_to_s(buf, buf_len, reg, context, value);
```

Reply:
OK. Let's just leave it as is.